pkg/filehandler/v1/file_handler.go

```diff
@@ -71,3 +86,183 @@ func (b BoltFileHandler) RemoveBucket(ctx context.Context, bucket string) error
+	return &FsFileHandler{Lock: &sync.Mutex{}, fileMap: make(map[string]*File), dirPath: path}, nil
+}
+
+func NormalizeBucketName(bucket string) string {
+	return strings.ReplaceAll(bucket, "/", "_")
+}
+
+func (f *FsFileHandler) AddFile(ctx context.Context, bucket, file string) error {
+	// Normalize the bucket name
+	bucket = NormalizeBucketName(bucket)
+
+	// Check if bucket exists in the map
+	f.Lock.Lock()
+	fileObject, ok := f.fileMap[bucket]
+	if !ok {
+		// If not, create a new file object and add it to the map
+		storageFileName := fmt.Sprintf("%s/%s.txt", f.dirPath, bucket)
+		fileOsObject, err := os.OpenFile(storageFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			f.Lock.Unlock()
+			return err
+		}
+		bufferedWriter := bufio.NewWriter(fileOsObject)
+		fileObject = &File{FileObject: fileOsObject, Lock: &sync.Mutex{}, Writer: bufferedWriter}
+		f.fileMap[bucket] = fileObject
+	}
+	f.Lock.Unlock()
+	// Lock the file
+	fileObject.Lock.Lock()
+	fileObject.FileObject.WriteString(file + "\n")
+	defer fileObject.Lock.Unlock()
+	return nil
+}
+
+func (f *FsFileHandler) Close() {
+	f.Lock.Lock()
+	for _, file := range f.fileMap {
+		file.Lock.Lock()
+		_ = file.FileObject.Close()
+	}
+}
+
+func (f *FsFileHandler) GetFiles(ctx context.Context, container string) (map[string]bool, error) {
+	// Normalize the bucket name
+	container = NormalizeBucketName(container)
+
+	// Create a map of string to bool
+	fileList := make(map[string]bool)
+	// Check if bucket exists in the map
+	f.Lock.Lock()
+	fileObject, ok := f.fileMap[container]
+	if !ok {
+		f.Lock.Unlock()
+		return fileList, fmt.Errorf("bucket does not exist for container %s", container)
+	}
+	f.Lock.Unlock()
+
+	// Lock the file
+	fileObject.Lock.Lock()
+	// Close the file
+	fileObject.FileObject.Close()
+
+	// Do not use defer here! It makes life much harder
+
+	// Copy the file to a new file
+	storageFileName := fmt.Sprintf("%s/%s.txt", f.dirPath, container)
+	storageFileCopyName := fmt.Sprintf("%s/%s_copy.txt", f.dirPath, container)
+	// Copy the file
+	err := copyFile(storageFileName, storageFileCopyName)
+	if err != nil {
+		fileObject.FileObject, _ = os.OpenFile(storageFileName, os.O_CREATE|os.O_WRONLY, 0644)
+		fileObject.Lock.Unlock()
+		return fileList, err
+	}
+
+	// Open the file
+	fileObject.FileObject, err = os.OpenFile(storageFileName, os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		fileObject.Lock.Unlock()
+		return fileList, err
+	}
+	fileObject.Writer = bufio.NewWriter(fileObject.FileObject)
+
+	// Unlock the file and let other goroutines use it
+	fileObject.Lock.Unlock()
+
+	// Read the file line by line
+	file, err := os.Open(storageFileCopyName)
+	if err != nil {
+		return fileList, err
+	}
+	defer file.Close()
+	defer os.Remove(storageFileCopyName)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		fileList[scanner.Text()] = true
+	}
+	if err := scanner.Err(); err != nil {
+		return fileList, err
+	}
+	return fileList, nil
+}
```

Comment:
Need to protect this as per bucket against re-entrancy (there is a single "copy" file that can be overwritten)